File: src/main/java/com/dingdongdeng/coinautotrading/trading/backtesting/context/TradingTimeContext.java
package com.dingdongdeng.coinautotrading.trading.backtesting.context;

import java.time.LocalDateTime;
import java.util.Objects;
import java.util.function.Supplier;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

public class TradingTimeContext {

    private static final ThreadLocal<Context> contextThreadLocal = ThreadLocal.withInitial(Context::new);

    public static LocalDateTime now() {
        Context context = contextThreadLocal.get();
        if (Objects.isNull(context)) {
            Context newContext = new Context();
            contextThreadLocal.set(newContext);
            return newContext.getNowSupplier().get();
        }
        return context.getNowSupplier().get();

    }

    public static void nowSupplier(Supplier<LocalDateTime> supplier) {
        Context context = contextThreadLocal.get();
        context.setNowSupplier(supplier);
    }

    public static void clear() {
        contextThreadLocal.remove();
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Context {

        private Supplier<LocalDateTime> nowSupplier = LocalDateTime::now;
    }

}

File: app-service/src/main/kotlin/com/dingdongdeng/autotrading/infra/common/utils/TradingTimeContext.kt
package com.dingdongdeng.autotrading.infra.common.utils

import java.time.LocalDateTime
import java.util.Objects

object TradingTimeContext {
    private val contextThreadLocal = ThreadLocal.withInitial { Context() }
    fun now(): LocalDateTime {
        val context = contextThreadLocal.get()
        if (Objects.isNull(context)) {
            val newContext = Context()
            contextThreadLocal.set(newContext)
            return newContext.now()
        }
        return context.now()
    }

    fun nowSupplier(supplier: () -> LocalDateTime) {
        val context = contextThreadLocal.get()
        context.now = supplier
    }

    fun clear() {
        contextThreadLocal.remove()
    }
}

class Context(
    var now: () -> LocalDateTime = { LocalDateTime.now() }
)

